core/src/main/java/org/jruby/parser/RubyParserBase.java

```diff
@@ -1511,28 +1513,30 @@ protected ArgsTailHolder new_args_tail(int line, ListNode keywordArg,
         return new_args_tail(line, keywordArg, keywordRestArgName, blockArg);
     }
 
-
-    @SuppressWarnings("CollectionIncompatibleType") // TODO: this should get reviewed/fixed, the error-prone warning is legit!
-    public Node remove_duplicate_keys(HashNode hash) {
-        List<Node> encounteredKeys = new ArrayList<>();
+    public Node remove_duplicate_keys(final HashNode hash) {
+        final Map<Node, KeyValuePair<Node, Node>> encounteredKeys = new HashMap<>();
+        List<KeyValuePair<Node, Node>> pairsToRemove = null;
 
         for (KeyValuePair<Node,Node> pair: hash.getPairs()) {
-            Node key = pair.getKey();
-            if (key == null || !(key instanceof LiteralValue)) continue;
-            int index = encounteredKeys.indexOf(key);
-            if (index >= 0) {
+            final Node key = pair.getKey();
+            if (!(key instanceof LiteralValue)) continue;
+            if (encounteredKeys.containsKey(key)) {
                 Ruby runtime = getConfiguration().getRuntime();
                 IRubyObject value = ((LiteralValue) key).literalValue(runtime);
                 warning(ID.AMBIGUOUS_ARGUMENT, lexer.getFile(), hash.getLine(), str(runtime, "key ", value.inspect(),
-                        " is duplicated and overwritten on line " + (encounteredKeys.get(index).getLine() + 1)));
-            } else {
-                encounteredKeys.add(key);
+                        " is duplicated and overwritten on line " + (key.getLine() + 1)));
+
+                if (pairsToRemove == null) pairsToRemove = new ArrayList<>(4);
+                pairsToRemove.add(encounteredKeys.get(key));
             }
+            // even if the key was previously seen, we replace the value to properly remove multiple duplicates
+            encounteredKeys.put(key, pair);
         }
 
-        for (Node key: encounteredKeys) {
-            hash.getPairs().remove(key);
-        }
+        // NOTE: we do not really remove the value part (RHS) as we still need to evaluate the expression - despite
+        // the value being dropped the side effects are desired and something MRI evaluates explicitly during removal
+        // if (pairsToRemove != null) hash.removeAll(pairsToRemove);
+
```

Comment:
I think you can just remove that removeAll comment and leave the text.  That code is not correct even if we do eventually address the comment.

Reply:
okay I considered removing but in that case I should do more cleanup - since we no longer need the `pairsToRemove` collection. having the comment without the collected nodes to be removed seemed weird, but I am pbly biased here :wink:

Reply:
@kares I know think this should not be in parsing at all but in building.  Building is capable of emitting instrs for the rhs and also processes kwargs in its' own right.  So you could just write a comment that IR Build is responsible for modifying the hash too?